pyiga/hierarchical.py

```diff
@@ -484,22 +541,64 @@ def deactivated_indices(self):
         """
         return self.ravel_indices(self.deactfun)
 
+    def _compute_single_axis_single_level_dirichlet_cells(self, lv, bdspec):
+        assert 0 <= lv < self.numlevels, 'Invalid level.'
+        return set(tuple(iter) for iter in assemble.boundary_cells(self.hmesh.meshes[lv].kvs, bdspec, ravel=False))
+
     def _compute_single_axis_single_level_dirichlet_indices(self, lv, bdspec):
         assert 0 <= lv < self.numlevels, 'Invalid level.'
-        bdax, bdside = bdspec
-        N = tuple(kv.numdofs for kv in self.hmesh.meshes[lv].kvs)
-        return set(tuple(iter) for iter in assemble.slice_indices(bdax, 0 if bdside==0 else -1, N, ravel=False))
+        return set(tuple(iter) for iter in assemble.boundary_dofs(self.hmesh.meshes[lv].kvs, bdspec, ravel=False))
+    
+    def _drop_bdspec_tuple_components(self, iterable, bdspec):
+        """Remove `bdspec`-components from `iterable` of tuples"""
+        if not iterable or not bdspec:
+            return set()
+        type_iterable = type(iterable)
+        out_components = list(range(self.dim))
+        out_components.pop(bdspec[0])
+        out = np.array(list(iterable.copy()))[:,out_components]
+        return type_iterable(tuple(iter) for iter in out)
 
     def _dirichlet_indices(self):
```

Comment:
There is no way to compute the boundary `HSpace` for a boundary which is not in the `bdspecs`. The computation of the boundary space should not be done in `_dirichlet_indices`, and the result should not be cached in the `HSpace` (it will become outdated when refinements are done). There should be a separate method `HSpace.boundary(bdspec)` which computes the boundary space for a single boundary.